src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,608 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.ConvertLinq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq
+{
+    internal sealed class CSharpConvertLinqQueryToLinqMethodProvider : AbstractConvertLinqProvider
+    {
+        private static readonly TypeSyntax VarNameIdentifier = SyntaxFactory.IdentifierName("var");
+
+        protected override IAnalyzer CreateAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+            => new CSharpAnalyzer(semanticModel, cancellationToken);
+
+        private sealed class CSharpAnalyzer : AnalyzerBase<QueryExpressionSyntax, SyntaxNode, SyntaxNode>
+        {
+            public CSharpAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+                : base(semanticModel, cancellationToken)
+            {
+            }
+
+            protected override string Title => CSharpFeaturesResources.Convert_LINQ_query_to_foreach;
+
+            protected override bool TryConvert(QueryExpressionSyntax source, out DocumentUpdate documentUpdate)
+            {
+                if (_semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any() ||
+                    source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                        SyntaxKind.SingleLineCommentTrivia,
+                        SyntaxKind.MultiLineCommentTrivia,
+                        SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                        source.ContainsDirectives))
+                {
+                    documentUpdate = default;
+                    return false;
+                }
+
+                return TryConvertInternal(source, source.Parent, out documentUpdate);
+            }
+
+            private bool TryConvertInternal(QueryExpressionSyntax source, SyntaxNode parent, out DocumentUpdate documentUpdate)
```

Comment:
is "parent" always source.Parent?  if so, arg may not be necessary.

Reply:
Definitely, not just the source.Parent. Please see
 case SyntaxKind.ParenthesizedExpression:
                        return TryConvertInternal(source, parent.Parent, out documentUpdate);

We may need to skip extra parenthesis.

---
In reply to: [176296929](https://github.com/dotnet/roslyn/pull/25362#discussion_r176296929) [](ancestors = 176296929)

Reply:
Note.  this approach to skipping parens seems very... baroque.  I think i'd much rather have a non-recursive algorithm that simply gets the parent, walking up parens if there are any.